cluster/distributedtask/scheduler.go

```diff
@@ -0,0 +1,365 @@
+
+			s.logger.WithFields(logrus.Fields{
+				"namespace":   namespace,
+				"taskID":      desc.ID,
+				"taskVersion": desc.Version,
+			}).Info("terminated distributed task execution")
+		}
+
+		// Check that all tasks that are already finished and if their TTL has passed, so we can clean them up.
+		cleanableTasks := filterTasks(tasks, func(task *Task) bool {
+			return task.Status != TaskStatusStarted && s.completedTaskTTL <= s.clock.Since(task.FinishedAt)
+		})
+		for _, task := range cleanableTasks {
+			err := s.taskCleaner.CleanUpDistributedTask(context.Background(), namespace, task.ID, task.Version)
+			if err != nil {
+				s.sampledLogger.WithSampling(func(l logrus.FieldLogger) {
+					l.WithFields(logrus.Fields{
+						"namespace":   namespace,
+						"taskID":      task.ID,
+						"taskVersion": task.Version,
+						"error":       err,
+					}).Error("failed to clean up distributed task")
+				})
+				continue
+			}
+
+			s.logger.WithFields(logrus.Fields{
+				"namespace":   namespace,
+				"taskID":      task.ID,
+				"taskVersion": task.Version,
+			}).Info("successfully submitted request to clean up distributed task")
+		}
+
+		// Check that tasks that can be cleaned up locally
+		localTasks := provider.GetLocalTasks()
+		for _, desc := range localTasks {
+			if _, ok := tasks[desc]; ok {
+				// task still present in the list
+				continue
+			}
+
+			if err := provider.CleanupTask(desc); err != nil {
+				s.sampledLogger.WithSampling(func(l logrus.FieldLogger) {
+					l.WithFields(logrus.Fields{
+						"namespace":   namespace,
+						"taskID":      desc.ID,
+						"taskVersion": desc.Version,
+						"error":       err,
+					}).Error("failed to clean up local distributed task state")
+				})
+			}
+		}
+	}
+}
+
+func (s *Scheduler) listTasks(ctx context.Context) (map[string]map[TaskDescriptor]*Task, error) {
+	tasksByNamespace, err := s.tasksLister.ListDistributedTasks(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "list distributed tasks")
+	}
+
+	result := make(map[string]map[TaskDescriptor]*Task, len(tasksByNamespace))
+	for namespace, tasks := range tasksByNamespace {
+		result[namespace] = make(map[TaskDescriptor]*Task, len(tasks))
+		for _, task := range tasks {
+			result[namespace][task.TaskDescriptor] = task
+		}
+	}
+	return result, nil
+}
+
+func (s *Scheduler) setRunningTaskHandleWithLock(namespace string, desc TaskDescriptor, handle TaskHandle) {
+	if _, ok := s.runningTasks[namespace]; !ok {
+		s.runningTasks[namespace] = map[TaskDescriptor]TaskHandle{}
+	}
+	s.runningTasks[namespace][desc] = handle
+}
+
+func (s *Scheduler) Close() {
+	close(s.stopCh)
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	for _, tasks := range s.runningTasks {
+		for _, task := range tasks {
+			task.Terminate()
+		}
+	}
+}
+
+func (s *Scheduler) totalRunningTaskCount() int {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	count := 0
+	for _, tasks := range s.runningTasks {
+		count += len(tasks)
+	}
```

Comment:
question: is this equal to `len(s.runningTasks)`?

Reply:
The `s.runningTasks` has a type of `map[string]map[string]*Task` (namespace -> taskID -> Task)

Reply:
aaa, sorry, I missed this, you're right